cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableRetryTest.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.spi.data.sql;
+
+import io.cdap.cdap.spi.data.transaction.TransactionException;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import org.mockito.Mockito;
+
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Test for SQL structured table with retries.
+ */
+public class SqlStructuredTableRetryTest extends SqlStructuredTableTest {
+  private static final String SQL_TRANSACTION_EXCEPTION = "SqlTransactionException";
+  private static final String SQL_CONNECTION_EXCEPTION = "SqlConnectionException";
+  private static final String TRANSACTION_CONFLICT_SQL_STATE = "40001";
+  private static final String CONNECTION_FAILURE_SQL_STATES = "08*";
+  private static final String SQL_TRANSACTION_EXCEPTION_MESSAGE = "Mocked Transaction Failure";
+  private static final String SQL_CONNECTION_EXCEPTION_MESSAGE = "Mocked Connection Failure";
+
+  @Override
+  protected TransactionRunner getTransactionRunner() throws TransactionException {
+    TransactionRunner transactionRunner = super.getTransactionRunner();
+    Set<Integer> failedCalls = new HashSet<>();
+    RetryingSqlTransactionRunner mockedRetryingSqlTransactionRunner =
+      (RetryingSqlTransactionRunner) Mockito.spy(transactionRunner);
+    SqlTransactionRunner mockedSqlTransactionRunner =
+      Mockito.spy(((RetryingSqlTransactionRunner) transactionRunner).getSqlTransactionRunner());
+
+    Mockito.doAnswer(i -> {
+      if (failedCalls.isEmpty()) {
+        // Fail first transaction with sql transaction failure.
+        failedCalls.add(1);
+        throw new SqlTransactionException(new SQLException(SQL_TRANSACTION_EXCEPTION, TRANSACTION_CONFLICT_SQL_STATE),
+                                          SQL_TRANSACTION_EXCEPTION_MESSAGE);
+      } else if (failedCalls.contains(1)) {
```

Comment:
I believe you don't need a counter.  You can just do 
```
Mockito.doAnswer(firstAnswer).doAnswer(secondAnswer).doAnswer(thirstAnswer).when...
```
see https://www.javadoc.io/doc/org.mockito/mockito-core/2.3.4/org/mockito/stubbing/Stubber.html.

Not sure if it would work for repeated scenarios (1,2,3,1,2,3,...). If so, I don't think you need the whole HashSet. It will alsways have 0/1 items. You can just use AtomitInteger or similar.

Reply:
Thanks for the pointer, updated.